File: algorithms/typescript/quick-sort.ts
function qs(arr: number[], lo: number, hi: number): void {
  if (lo >= hi) {
    return;
  }
  const pivotIdx = partition(arr, lo, hi);
  qs(arr, lo, pivotIdx - 1);
  qs(arr, pivotIdx + 1, hi);
}
function partition(arr: number[], lo: number, hi: number): number {
  const pivot = arr[hi];

  let idx = lo - 1;
  for (let i = lo; i < hi; i++) {
    if (arr[i] <= pivot) {
      idx++;
      const temp = arr[i];
      arr[i] = arr[idx];
      arr[idx] = temp;
    }
  }
  idx++;
  arr[hi] = arr[idx];
  arr[idx] = pivot;
  return idx;
}
function quick_sort(arr: number[]): void {
  qs(arr, 0, arr.length - 1);
}

File: algorithms/python/quick-sort.py
def qs(arr: list[int], lo: int, hi: int):
    if lo >= hi:
        return
    pivotIdx = partition(arr, lo, hi)
    qs(arr, lo, pivotIdx - 1)
    qs(arr, pivotIdx + 1, hi)


def partition(arr: list[int], lo: int, hi: int) -> int:
    pivot = arr[hi]
    idx = lo - 1
    for i in range(lo, hi):
        if arr[i] <= arr[pivot]:
            idx += 1
            arr[i], arr[idx] = arr[idx], arr[i]
    idx += 1
    arr[idx], pivot = pivot, arr[idx]
    return idx


def quick_sort(arr: list[int]):
    qs(arr, 0, len(arr) - 1)
